### PR TITLE
[fix] #2123: Return to `PublicKey` de/serialization from multihash

### LINKED ISSUE #2123: Use enum for `PublicKey`/`PrivateKey` digest function
Both `PublicKey` and `PrivateKey` store their respective digest functions as a string. They should store it as an `Algorithm` enum provided in the `iroha_crypto` crate.

### PR DESCRIPTION
**Pull request by Erigara**
### Description of the Change

Return `PublicKey` to de/serialization from multihash string.

### Issue

Closes #2123.

### Benefits

Reverse breaking change.

### Possible Drawbacks

None.

### CODE REVIEW

File: crypto/src/lib.rs [hunk 1]
```diff
@@ -403,6 +328,25 @@ impl KeyPair {
     }
 }
 
+#[cfg(feature = "std")]
```
Comment (appetrosyan) [inline crypto/src/lib.rs:331]: This is an interesting approach. The most readable `impl<'de> Deserialize<'de>` which wasn't redundant. I wish we used this approach more.

**Review by appetrosyan — DISMISSED** [review]

**Review by pesterev — APPROVED** [review]
LGTM

File: crypto/Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,7 @@ iroha_schema = { path = "../schema" }
 derive_more = { version = "0.99.17", default-features = false, features = ["deref", "deref_mut", "display"] }
 parity-scale-codec = { version = "3.2.1", default-features = false, features = ["derive", "full"] }
 serde = { version = "1.0.151", default-features = false, features = ["derive"] }
+serde_with = {version = "2.1.0", default-features = false, features = ["macros"]}
```
Comment (Arjentix) [inline crypto/Cargo.toml:30]:
```suggestion
serde_with = { version = "2.1.0", default-features = false, features = ["macros"] }
```

File: crypto/src/lib.rs [hunk 3]
```diff
@@ -267,7 +192,7 @@ pub enum Error {
     #[display(fmt = "Signing failed. {_0}")]
     Signing(String),
     /// Returned when an error occurs during key generation
-    #[display(fmt = "Key generation failed. {_0} sss")]
```
Comment (Arjentix) [inline crypto/src/lib.rs:270]: Lol

File: crypto/src/lib.rs [hunk 4]
```diff
@@ -355,7 +280,7 @@ impl KeyPair {
 
                 SignatureOf::new(key_pair.clone(), &dummy_payload)?
                     .verify(&dummy_payload)
-                    .map_err(|_err| Error::KeyGen(String::from("Key pair mismatch sss")))?;
```
Comment (Arjentix) [inline crypto/src/lib.rs:358]: Lol x2

File: crypto/src/lib.rs [hunk 5]
```diff
@@ -403,6 +328,25 @@ impl KeyPair {
...
+        #[derive(Deserialize)]
+        struct KeyPair {
+            public_key: PublicKey,
+            private_key: PrivateKey,
+        }
+
+        let key_pair = KeyPair::deserialize(deserializer)?;
+        Self::new(key_pair.public_key, key_pair.private_key).map_err(D::Error::custom)
+    }
+}
```
Comment (Arjentix) [inline crypto/src/lib.rs:348]: I don't get it. Original `KeyPair` has the exact same fields. What does this custom impl does that can't be done with `Deserialize` derive?

Comment (Arjentix) [reply crypto/src/lib.rs:348]: To future readers. Roman answered in DM that it's done to use the `new()` method to check keys consistency.

In that case I suggest to note this in the code and/or comments. I.e. we can remove internal `KeyPair` to `KeyPairUnchecked`.

Comment (appetrosyan) [reply crypto/src/lib.rs:348]: I thought it was obvious from the signature. IMO KeyPairCandidate is a bit better, and a comment just above ::new to emphasise the consistency check would help.

File: crypto/Cargo.toml (hunk 2 above)
Comment (appetrosyan) [reply crypto/Cargo.toml:30]: On that note, I feel we should consider adding a TOML auto-formatter to the CI as well as cargo fmt. In most cases the overhead would be minimal.

Comment (appetrosyan) [reply crypto/Cargo.toml:30]: #3282

File: crypto/src/lib.rs [hunk 6]
```diff
@@ -71,101 +72,25 @@ impl std::error::Error for NoSuchAlgorithm {}
 
 ffi::ffi_item! {
     /// Algorithm for hashing
-    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, FfiType, IntoSchema)]
+    #[derive(Debug, Display, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, FfiType, IntoSchema, SerializeDisplay, DeserializeFromStr, Encode, Decode, Default)]
```
[the remark this answers is not in the document]
Comment (Erigara) [reply crypto/src/lib.rs:75]: Sure, i will change this

**Review by Arjentix — APPROVED** [review]

**Review by pesterev — APPROVED** [review]

**Pull request merged**